scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/DownloadDatasetController.java

```diff
@@ -0,0 +1,82 @@
+package edu.asu.diging.scriptoocloud.web;
+
+import edu.asu.diging.scriptoocloud.core.exceptions.FileSystemStorageException;
+import edu.asu.diging.scriptoocloud.core.model.impl.Dataset;
+import edu.asu.diging.scriptoocloud.core.service.IDatasetService;
+import edu.asu.diging.scriptoocloud.core.service.IFileSystemService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.*;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.*;
+import java.security.Principal;
+import java.util.Locale;
+
+@Controller
+public class DownloadDatasetController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final IDatasetService iDatasetService;
+    private final IFileSystemService iFileSystemService;
+
+    @Autowired
+    public DownloadDatasetController(IDatasetService iDatasetService, IFileSystemService iFileSystemService) {
+        this.iDatasetService = iDatasetService;
+        this.iFileSystemService = iFileSystemService;
+    }
+
+    @RequestMapping(value = "auth/datasets/{id}/download")
+    @PreAuthorize("hasPermission(#datasetId, 'Dataset', 'download')")
+    public void download(@PathVariable("id") Long datasetId,
+                         HttpServletResponse response,
+                         Principal principal) throws IOException, FileSystemStorageException {
+        String username = principal.getName();
+        String type = Dataset.class.getSimpleName().toLowerCase(Locale.ROOT);
+        Dataset dataset = (Dataset) iDatasetService.findById(datasetId);
+        if (dataset != null) {
+            String filename = type + "_" + datasetId + "_version_" + dataset.getVersion() + ".zip";
+            try {
+                iFileSystemService.createZipFile(username, type, datasetId.toString(),
```

Comment:
I would create the zip file asynchronically. This can take a long time if there are many files.